lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -430,6 +430,8 @@ class PullbackCloner::Implementation final
     }
     case AdjointValueKind::AddElement: {
       auto baseAdjoint = val;
+      assert(baseAdjoint.getType().is<TupleType>() ||
```

Comment:
This is not what should be checked here. This assertion is enforced during `AddElement` construction. Here you need to ensure that all nested adjoints are of the same kind: either all tuples or all structs, but not the mixture.

Reply:
I see. Sorry I sent out that last revision in haste. Fixed in the latest revision and squashed commits.